File: almoosa_customization/almoosa_customization/report/sales_summary_per_group/sales_summary_per_group.py
import frappe
from collections import defaultdict

def execute(filters=None):
    columns = []
    data = []

    conditions = ""
    values = {}

    # -----------------------
    # Date & Time Filter
    # -----------------------
    if filters.get("from_datetime") and filters.get("to_datetime"):
        conditions += """
            AND TIMESTAMP(hd.posting_date, hd.posting_time)
            BETWEEN %(from_datetime)s AND %(to_datetime)s
        """
        values["from_datetime"] = filters.get("from_datetime")
        values["to_datetime"] = filters.get("to_datetime")

    # -----------------------
    # Item Group Filter
    # -----------------------
    if filters.get("item_group"):
        conditions += """
            AND SUBSTRING_INDEX(it.item_group, '.', -1) IN %(item_group)s
        """
        values["item_group"] = tuple(filters.get("item_group"))

    # -----------------------
    # Warehouse Filter
    # -----------------------
    if filters.get("warehouse"):
        conditions += " AND it.warehouse IN %(warehouse)s"
        values["warehouse"] = tuple(filters.get("warehouse"))

    # -----------------------
    # Fetch Data
    # -----------------------
    records = frappe.db.sql(f"""
        SELECT 
            SUBSTRING_INDEX(it.item_group, '.', -1) AS item_group,
            LEFT(it.warehouse,3) warehouse,
            SUM(it.qty) AS qty
        FROM `tabPOS Invoice Item` it
        INNER JOIN `tabPOS Invoice` hd 
            ON hd.name = it.parent
        WHERE hd.docstatus = 1 {conditions}
        GROUP BY it.item_group, it.warehouse
        ORDER BY item_group
    """, values, as_dict=True)

    # -----------------------
    # Prepare Pivot Structure
    # -----------------------
    item_group_data = defaultdict(lambda: defaultdict(float))
    warehouses = set()

    for row in records:
        item_group = row["item_group"]
        warehouse = row["warehouse"]
        qty = row["qty"] or 0

        item_group_data[item_group][warehouse] += qty
        warehouses.add(warehouse)

    warehouses = sorted(warehouses)

    # -----------------------
    # Columns
    # -----------------------
    columns.append({
        "label": "Item Group",
        "fieldname": "item_group",
        "fieldtype": "Data",
        "width": 130
    })

    for wh in warehouses:
        columns.append({
            "label": wh,
            "fieldname": wh,
            "fieldtype": "Int",
            "width": 60
        })

    columns.append({
        "label": "Total",
        "fieldname": "total",
        "fieldtype": "Int",
        "width": 100
    })

    # -----------------------
    # Data Rows
    # -----------------------
    grand_totals = defaultdict(float)

    for item_group in sorted(item_group_data.keys()):
        row = {"item_group": item_group}
        total = 0

        for wh in warehouses:
            qty = item_group_data[item_group].get(wh, 0)
            row[wh] = qty
            total += qty
            grand_totals[wh] += qty

        row["total"] = total
        data.append(row)

    # -----------------------
    # Grand Total Row
    # -----------------------
    if data:
        grand_row = {"item_group": "Grand Total"}
        grand_total_sum = 0

        for wh in warehouses:
            grand_row[wh] = grand_totals[wh]
            grand_total_sum += grand_totals[wh]

        grand_row["total"] = grand_total_sum
        data.append(grand_row)

    return columns, data

File: almoosa_customization/almoosa_customization/report/sales_summary_per_group/sales_summary_per_group.js
// Copyright (c) 2026, Printechs and contributors
// For license information, please see license.txt

frappe.query_reports["Sales Summary Per Group"] = {
	"filters": [
		{
            "fieldname": "from_datetime",
            "label": "From Date & Time",
            "fieldtype": "Datetime",
            "default": frappe.datetime.month_start() + " 00:00:01"
        },

        {
            "fieldname": "to_datetime",
            "label": "To Date & Time",
            "fieldtype": "Datetime",
            "default": frappe.datetime.month_end() + " 23:59:59"
        },
		{
			"fieldname": "item_group",
			"label": "Item Group",
			"fieldtype": "MultiSelectList",
			"options": "Item Group",
			"get_data": function(txt) {
				return frappe.db.get_link_options("Item Group", txt).then(function(data) {
					return data.map(function(d) {
						return {
							value: d.value,   // full name (important for backend)
							description: d.value.split('.').pop()  // last part only
						};
					});
				});
			}
		},

		{
			"fieldname": "warehouse",
			"label": "Warehouse",
			"fieldtype": "MultiSelectList",
			"options": "Warehouse",
			"get_data": function(txt) {
				return frappe.db.get_link_options("Warehouse", txt);
			}
		},

	],
	
	"formatter": function (value, row, column, data, default_formatter) {

        value = default_formatter(value, row, column, data);

        if (!data) {
            return value;
        }

        // ✅ Bold TOTAL column (all rows)
        if (column.fieldname === "total") {
            value = `<span style="font-weight:600;">${value}</span>`;
        }

        // ✅ Bold GRAND TOTAL row
        if (data.item_group === "Grand Total") {
            value = `<span style="font-weight:700;">${value}</span>`;
        }

        return value;
}
};
